subsys/bluetooth/mesh/friend.c

```diff
@@ -1676,7 +1676,8 @@ static bool friend_queue_prepare_space(struct bt_mesh_friend *frnd, uint16_t add
 	pending_segments = false;
 
 	while (pending_segments || avail_space < seg_count) {
-		struct net_buf *buf = (void *)net_buf_slist_get(&frnd->queue);
+		struct net_buf *buf =
+			CONTAINER_OF(sys_slist_get(&frnd->queue), struct net_buf, node);
```

Comment:
This makes me a little bit uneasy, since it's not quite the same as the original code. The `node` struct member happens to be the first one in `struct net_buf`, i.e. its address coincides with the address of the struct, however using `CONTAINER_OF()` makes it look like it would now be safe to move `node` to some other position within the struct. This is not the true however, since you're doing the NULL check: the slist API could return NULL, but then `CONTAINER_OF()` would convert the pointer to something non-NULL, making it seem like you got a valid pointer when in fact you didn't.

Reply:
That's a pretty good point, didn't think of that. Should I change the relevant locations to get a `sys_snode_t`, and only use `CONTAINER_OF` later after the NULL check?

For example, in another location:

```c
/* Discard queued buffers */
while ((buf = CONTAINER_OF(sys_slist_get(&att->prep_queue), struct net_buf, node))) {
	net_buf_unref(buf);
}
```
to
```c
/* Discard queued buffers */
while ((node = sys_slist_get(&att->prep_queue))) {
	net_buf_unref(CONTAINER_OF(node, struct net_buf, node));
}
```
?

Reply:
Just FYI, both loops you've suggested would break rule 79 of https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html I think. 

You probably need to use a `do { } while` instead

Reply:
@kamnxt your proposal looks ok to me. As for the comment from @Thalley, what should hopefully keep the code size the same, would be to use `sys_slist_is_empty()` as the while-loop condition, after which you can use `sys_slist_get_not_empty()` inside the loop.

Reply:
Btw, with the separate `!sys_slist_is_emppty()` check, the `CONTAINER_OF()` actually does become safe since it'd never be called on a NULL pointer.

Reply:
@Thalley the snippets in my comment were based on existing code in [att.c](https://github.com/zephyrproject-rtos/zephyr/pull/87577/files#diff-3adc165d775d407a3eb6b90da365671eee38ca4e38d9b6f6661abf2975a5161e)...

@jhedberg your suggestion seems nice, will push a change soon

Reply:
@jhedberg @Thalley better?

Reply:
Looks better to me at least. I haven't looked enough into it to determine if it's done in a safe manner; doing a `sys_slist_is_empty` followed by a `sys_slist_get_not_empty` doesn't sound thread safe, so only if a single thread is ever accessing the lists can this be considered safe. 

Maybe @jhedberg or @alwa-nordic know more about whether the lists changed in Bluetooth are accessed by multiple threads

Reply:
> doing a sys_slist_is_empty followed by a sys_slist_get_not_empty doesn't sound thread safe, so only if a single thread is ever accessing the lists can this be considered safe

This is just as safe/unsafe as a single `sys_slist_get()` call. The slist API is documented as not being multithread safe. Since the old net_buf_slist API did have a lock assume @kamnxt has done some level of analysis and concluded that locking is not needed?

Reply:
@kamnxt note that in most places locking in the Bluetooth stack is implicit, since almost everything runs in cooperative threads. The places that need protection are where a preemptible app-side thread might call into the stack or some shared data might be accessed in ISR context.

Reply:
I'm fairly confident we rely on a thread safe `net_buf_slist_get`, so we need to make sure a refactor does the right thing.

@kamnxt I think we regard `sys_slist_is_empty` as thread safe, even if not explicitly documented. The current implementation is thread safe (at least on architectures we care about).